### PR TITLE
fix(schema): support all AliasChoices bindings

### LINKED ISSUE #673: fix(schema): loader has no AliasChoices support — multi-alias fields misbind across all validate checks
Found while triaging Greptile threads on #667 and #672 (pre-existing; reviewer-verified).

`SchemaLoader.extract_metadata` records `alias=None` for a field whose `validation_alias` is a
non-string (`AliasChoices`/`AliasPath`), so the field binds by attribute name only:

```python
class M(BaseSettings):
    api_key: str = Field(validation_alias=AliasChoices("API_KEY", "LEGACY_API_KEY"),
                         json_schema_extra={"sensitive": True})
```

- A `.env` using `LEGACY_API_KEY=` is reported missing_required + extra_vars (wrong in both
  directions — pydantic-settings binds it fine).
- `API_KEY=` only matched by the accident of case-insensitive name coincidence (and #672's
  env_prefix work exposes that fragility: a prefixed model would bind `MYAPP_api_key` instead).
- The sensitive-suppression set holds ONE binding key per field, so secrets bound via an
  alternate choice still get contradictory "not marked sensitive" warnings (the #667 Greptile
  thread).

Ask: model multi-name bindings properly — either FieldMetadata carries the ordered list of
binding candidates (first-match-wins like pydantic-settings) consumed by _field_lookup_key and
every check (missing/extra/sensitive/type/constraints), or the loader expands to one canonical
binding plus accepted-alternates. Regression tests: each choice binds; suppression covers all
choices; prefix interaction (aliases bypass env_prefix); extra="forbid" doesn't flag an
alternate choice.

#672 ships the minimal slice (first string choice as effective alias so explicit aliases are
never silently discarded); this issue covers full multi-choice support.

### PR DESCRIPTION
**Pull request by jainal09**
## What was broken

For a field using `AliasChoices("API_KEY", "LEGACY_API_KEY")`, schema metadata
kept only `API_KEY`. A deterministic baseline run with only
`LEGACY_API_KEY=ordinary` exited 1 and reported all three incorrect outcomes:

- `api_key` was missing.
- `LEGACY_API_KEY` was extra under `extra="forbid"`.
- `LEGACY_API_KEY` was not marked sensitive.

The same pinned pydantic-settings 2.14.2 model binds the later choice successfully.
Scratch verification also confirmed that aliases bypass `env_prefix`, matching is
case-insensitive by default, `case_sensitive=True` requires exact casing, and the
first declared choice wins when several are present.

## What changed

- Add ordered `FieldMetadata.binding_names` while retaining `alias` and `env_name`
  as backward-compatible canonical values.
- Recognize every binding candidate for required/optional presence, extra-variable
  checks, sensitive suppression, and unencrypted-secret reporting.
- Select the first present candidate for base-type and constraint validation, in
  pydantic-settings order. Constraint input remains keyed by model aliases/names,
  not environment candidates.
- Preserve `env_prefix` behavior: plain fields use it and explicit aliases bypass
  it.
- Document multi-alias behavior in the validate and API references.

AliasPath keeps the existing behavior of using its first string component as the
environment binding. Traversing later path components such as integer indices is
out of scope for this change.

## Regression tests

Coverage includes:

- metadata for plain, string-alias, and ordered AliasChoices bindings;
- first and later choices binding independently;
- two choices present with first-choice-wins value selection;
- `extra="forbid"` accepting every declared choice;
- later-choice base-type and constraint failures;
- required and optional presence checks;
- sensitive suppression and unencrypted-secret reporting across all choices;
- aliases bypassing `env_prefix`; and
- exact-case AliasChoices matching.

Failing-before proof used `git checkout origin/main -- src/` and the focused run
failed with 11 failures and 6 passes. After the implementation, the expanded
focused run passed all 18 tests.

The deterministic CLI reproduction now exits 0, reports no missing/extra variable,
and emits only the expected unencrypted-sensitive warning.

## Review-round verification

- **Plain two-string choices — no bug reproduced.** Pinned Pydantic 2.12.5 accepted
  both `model_validate({"CURRENT_PORT": "0"})` and
  `model_validate({"LEGACY_PORT": "0"})`; each produced
  `greater_than_equal` at the supplied alias. The existing envdrift later-choice
  constraint test also failed correctly, so this case needed no case-specific fix.
- **Mixed AliasPath/string choices — fixed.** For
  `AliasChoices(AliasPath("NESTED_PORT", "value"), "LEGACY_PORT")` with only
  `LEGACY_PORT=0`, real dotenv loading produced `greater_than_equal` at
  `LEGACY_PORT`, while envdrift assembled `{"NESTED_PORT": "0"}` and falsely
  returned valid. The constraint pass now preserves the actually bound validation
  key, assembles `{"LEGACY_PORT": "0"}`, and reports the same constraint failure.
  A regression test directly compares real Pydantic with envdrift.
- **CodeScene extraction.** Per-field assembly moved to
  `_extract_field_metadata`; `SchemaLoader.extract_metadata` dropped from 116 to
  68 lines, making this PR neutral-to-improving for the pre-existing hotspot.
- **CodeScene test duplication.** The flagged tests remain separate because their
  outcome contracts differ: one pins successful binding and first-choice
  precedence, while the other pins base-type and constraint rejection shapes.

## Gates

- `uv sync --all-extras`
- `uv run ruff check src tests`
- `uv run ruff format --check .`
- `uv run pyrefly check`
- `uv run bandit -r src -c pyproject.toml`
- `uv run pytest -m "not integration"` — 3506 passed, 6 skipped, 563 deselected
- `env npm_config_cache=/home/jainal09/.cache/npm make lint-docs` — 0 errors

Closes #673

🤖 Generated with Codex (gpt-5.6-sol) orchestrated by Claude Code

<a href="https://cubic.dev/pr/jainal09/envdrift/pull/678?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved environment variable resolution for fields with validation aliases, `AliasChoices`, and prefixed bindings, including correct ordered “first-choice wins” behavior.
  * Fixed missing-variable, type/constraint, and sensitive/encryption checks to use the same multi-candidate binding logic (including accurate handling of encrypted vs plaintext and `env_ignore_empty`).
  * Enhanced `case_sensitive=True` matching so only exact-cased keys bind; incorrect casing is treated as extra/missing appropriately.
* **Documentation**
  * Clarified CLI and Validator API behavior for `env_prefix`, aliases, `AliasChoices`, case sensitivity, and how sensitive-field/encryption reporting works.
* **Tests**
  * Expanded coverage for ordered binding names and alias/prefix/encryption edge cases.

<h3>Greptile Summary</h3>

This PR expands schema validation to handle every `AliasChoices` environment binding. The main changes are:

- Adds ordered binding metadata while keeping legacy alias fields.
- Uses all binding candidates for presence, extra-variable, type, constraint, and sensitivity checks.
- Preserves first-choice-wins value selection and `env_prefix` bypass for explicit aliases.
- Documents multi-alias behavior in CLI and API references.
- Adds regression tests for later choices, mixed aliases, case sensitivity, and sensitive fields.

<h3>Confidence Score: 5/5</h3>

This looks safe to merge.

- No blocking issues found in the changed code.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/envdrift/core/schema.py | Adds ordered environment binding metadata for plain fields, aliases, and `AliasChoices`. |
| src/envdrift/core/validator.py | Updates validation paths to recognize all declared binding candidates. |
| tests/unit/test_schema_loader.py | Adds metadata coverage for ordered binding names. |
| tests/unit/test_validator.py | Adds regression coverage for multi-alias binding and validation behavior. |
| docs/cli/validate.md | Documents CLI validation behavior for `AliasChoices` bindings. |
| docs/reference/api.md | Documents API validation behavior for multi-alias environment bindings. |

<sub>Reviews (3): Last reviewed commit: ["Merge branch &#39;main&#39; into fix/schema-alia..."](https://github.com/jainal09/envdrift/commit/a7d9b2585b4c4caf34df6ca0f9d774b4390ba694) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=44113890)</sub>

### CODE REVIEW

File: src/envdrift/core/validator.py [hunk 1]
```diff
@@ -371,22 +400,26 @@ def _check_sensitive_encryption(
...
         # Also check for suspicious plaintext values. The dotenvx public-key
         # artifact is, by definition, public — its ``*_KEY`` name must not
         # produce a bogus "mark it sensitive" warning (#472).
         sensitive_names = {
-            _field_lookup_key(field_meta, schema.case_sensitive)
+            lookup_key
             for field_meta in schema.fields.values()
             if field_meta.sensitive
+            for lookup_key in _field_lookup_keys(field_meta, schema.case_sensitive)
         }
```
[the remark this answers is not in the document]
Comment (jainal09) [reply src/envdrift/core/validator.py:446]: Dispositioning (reviewer): this requires a model where a sensitive field and a non-sensitive field share the same alias choice — a pathological schema in which pydantic binds BOTH fields from one env var. The suppressed warning is the advisory 'looks like a secret but is not marked sensitive' nudge, and for that env name the nudge IS satisfied: the name is sensitively bound (the sensitive field's detection and unencrypted_secrets reporting are unaffected). Suppressing a duplicate advisory for a name the schema already marks sensitive via one of its bindings is the intended semantics of the suppression set; warning anyway would re-introduce the #658-style contradiction for every legitimately shared binding. If a real-world schema shape surfaces where this hides an actionable risk, happy to revisit with a repro. Resolving.

File: tests/unit/test_validator.py [hunk 2]
```diff
@@ -660,15 +660,46 @@ class TestValidatorEnvPrefix:
     """#669: validation mirrors pydantic-settings ``env_prefix`` bindings."""
 
     @pytest.mark.parametrize(
-        ("settings_cls", "env_name", "expected_alias", "expected_env_name"),
+        (
```
Comment (jainal09) [inline tests/unit/test_validator.py:663]: Dispositioning: 620 vs 600 lines in the subject's single test module — the marginal overage comes from the alias-binding regression matrix this PR adds, each test pinning a distinct binding claim. The one-module-per-subject layout is the repo's deliberate convention (dispositioned repeatedly on this module). Resolving.

File: tests/unit/test_validator.py [hunk 3]
```diff
@@ -1,7 +1,7 @@
 """Tests for Validator."""
 
 import pytest
-from pydantic import AliasChoices, Field, model_validator
+from pydantic import AliasChoices, AliasPath, Field, ValidationError, model_validator
```
[the remark this answers is not in the document]
Comment (jainal09) [reply tests/unit/test_validator.py:4]: Re-raise of the dispositioned LoC thread on the new head: the marginal overage is the alias-binding regression matrix (now including the mixed-AliasPath false-PASS regression test added at the reviewer's request this round). One-module-per-subject stands. Resolving.

[the remark this answers is not in the document]
Comment (jainal09) [reply tests/unit/test_validator.py:4]: Dispositioning: the two tests assert DIFFERENT outcome shapes — binding-order precedence (both choices present, first wins, validation passes with the bound value) versus type/constraint rejection through a later choice (validation fails). Parametrizing across pass-vs-fail outcomes needs branching assertions, which reads worse than named claims; same disposition as the equivalent pairs on #661/#672. Resolving.

**Comment by jainal09** [comment]
@coderabbitai full review
